File: motor3d 2.0.ConvertedToC#/Cara.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
using Motor3D.Espacio3D;
using Motor3D.Espacio2D;
using Motor3D.Escena;
using Motor3D.Escena.Shading;
using System.Drawing;
using System.Math;

namespace Motor3D.Primitivas3D
{
	public class Cara : ObjetoPrimitiva3D
	{

		private Vector3D mNormalSUR;
		private Vector3D mNormalSRC;
		private int[] mVertices;
		private Punto3D mBaricentroSUR;
		private Punto3D mBaricentroSRC;
		private Color mColor;
		private Color mColorShading;

		private bool mCargadaEnBuffer;
		public new event ModificadoEventHandler Modificado;
		public new delegate void ModificadoEventHandler(ref Cara Sebder);

		public bool CargadaEnBuffer {
			get { return mCargadaEnBuffer; }
			set { mCargadaEnBuffer = value; }
		}

		public Vector3D NormalSUR {
			get { return mNormalSUR; }
			set {
				mNormalSUR = value;
				if (Modificado != null) {
					Modificado(this);
				}
			}
		}

		public Vector3D NormalSRC {
			get { return mNormalSRC; }
			set {
				mNormalSRC = value;
				if (Modificado != null) {
					Modificado(this);
				}
			}
		}

		public Punto3D BaricentroSUR {
			get { return mBaricentroSUR; }
			set {
				mBaricentroSUR = value;
				if (Modificado != null) {
					Modificado(this);
				}
			}
		}

		public Punto3D BaricentroSRC {
			get { return mBaricentroSRC; }
			set {
				mBaricentroSRC = value;
				if (Modificado != null) {
					Modificado(this);
				}
			}
		}

		public int[] Vertices {
			get { return mVertices; }
			set {
				if (value.GetUpperBound(0) == mVertices.GetUpperBound(0)) {
					mVertices = value;
					if (Modificado != null) {
						Modificado(this);
					}
				}
			}
		}

		public int Vertices {
			get {
				if (Indice >= 0 && Indice <= mVertices.GetUpperBound(0)) {
					return mVertices[Indice];
				} else {
					return -1;
				}
			}
			set {
				if (Indice >= 0 && Indice <= mVertices.GetUpperBound(0)) {
					mVertices[Indice] = value;
					if (Modificado != null) {
						Modificado(this);
					}
				}
			}
		}

		public Vertice[] Vertices {
			get {
				Vertice[] Retorno = new Vertice[mVertices.GetUpperBound(0) + 1];

				for (int i = 0; i <= mVertices.GetUpperBound(0); i++) {
					Retorno[i] = ArrayVertices[mVertices[i]];
				}

				return Retorno;
			}
		}

		public Punto3D[] PuntosSUR {
			get {
				Punto3D[] Retorno = new Punto3D[mVertices.GetUpperBound(0) + 1];

				for (int i = 0; i <= mVertices.GetUpperBound(0); i++) {
					Retorno[i] = Vertices[mVertices[i]].CoodenadasSUR;
				}

				return Retorno;
			}
		}

		public Punto3D[] PuntosSRC {
			get {
				Punto3D[] Retorno = new Punto3D[mVertices.GetUpperBound(0) + 1];

				for (int i = 0; i <= mVertices.GetUpperBound(0); i++) {
					Retorno[i] = Vertices[mVertices[i]].CoodenadasSRC;
				}

				return Retorno;
			}
		}

		public Punto2D[] Representacion {
			get {
				Punto2D[] Retorno = new Punto2D[mVertices.GetUpperBound(0) + 1];

				for (int i = 0; i <= mVertices.GetUpperBound(0); i++) {
					Retorno[i] = Vertices[mVertices[i]].Representacion;
				}

				return Retorno;
			}
		}

		public Point[] RepresentacionToPoint {
			get {
				Point[] Retorno = new Point[mVertices.GetUpperBound(0) + 1];

				for (int i = 0; i <= mVertices.GetUpperBound(0); i++) {
					Retorno[i] = Vertices[mVertices[i]].Representacion.ToPoint();
				}

				return Retorno;
			}
		}

		public int NumeroVertices {
			get { return mVertices.GetUpperBound(0) + 1; }
		}

		public Color Color {
			get { return mColor; }
			set {
				mColor = value;
				if (Modificado != null) {
					Modificado(this);
				}
			}
		}

		public Color ColorShading {
			get { return mColorShading; }
			set { mColorShading = value; }
		}

		public Cara(params int[] Vertices)
		{
			if (Vertices.GetUpperBound(0) >= 2) {
				mVertices = Vertices;
				mNormalSUR = new Vector3D();
				mBaricentroSUR = new Punto3D();
				mColor = Color.White;
			} else {
				if (Vertices.GetUpperBound(0) == 0) {
					mVertices = new int[Vertices[0]];
					mNormalSUR = new Vector3D();
					mBaricentroSUR = new Punto3D();
					mColor = Color.White;
				} else {
					throw new ExcepcionPrimitiva3D("CARA (NEW): Una cara debe tener al menos tres vértices." + Constants.vbNewLine + "Número de vértices: " + Vertices.GetUpperBound(0) + 1);
				}

			}
		}

		public void RecalcularNormalSUR(Vertice[] Vertices)
		{
			mNormalSUR = Cara.VectorNormalSUR(this, Vertices);
			if (Modificado != null) {
				Modificado(this);
			}
		}

		public void RecalcularBaricentroSUR(Vertice[] Vertices)
		{
			mBaricentroSUR = Cara.BaricentroCaraSUR(this, Vertices);
			if (Modificado != null) {
				Modificado(this);
			}
		}

		public void RecalcularDatosSUR(Vertice[] Vertices)
		{
			mNormalSUR = Cara.VectorNormalSUR(this, Vertices);
			mBaricentroSUR = Cara.BaricentroCaraSUR(this, Vertices);
			if (Modificado != null) {
				Modificado(this);
			}
		}

		public void RecalcularNormalSRC(Vertice[] Vertices)
		{
			mNormalSRC = Cara.VectorNormalSRC(this, Vertices);
			if (Modificado != null) {
				Modificado(this);
			}
		}

		public void RecalcularBaricentroSRC(Vertice[] Vertices)
		{
			mBaricentroSRC = Cara.BaricentroCaraSRC(this, Vertices);
			if (Modificado != null) {
				Modificado(this);
			}
		}

		public void RecalcularDatosSRC(Vertice[] Vertices)
		{
			mNormalSRC = Cara.VectorNormalSRC(this, Vertices);
			mBaricentroSRC = Cara.BaricentroCaraSRC(this, Vertices);
			if (Modificado != null) {
				Modificado(this);
			}
		}

		public void Shading(PhongShader Constantes, Foco3D[] Focos, Camara3D Camara)
		{
			mColorShading = Constantes.EcuacionPhong(Focos, mNormalSUR, mBaricentroSUR, mColor, Camara);
		}

		public void RevertirVertices()
		{
			Array.Reverse(mVertices);
		}

		public bool EsVisible(Camara3D Camara, Vertice[] Vertices)
		{
			if (mNormalSRC.Z <= 0.01 && Camara.Frustum.Pertenece(mBaricentroSRC)) {
				for (int i = 0; i <= mVertices.GetUpperBound(0); i++) {
					if (Camara.Pantalla.Pertenece(Vertices[mVertices[i]].Representacion)) {
						return true;
					}
				}

				return false;
			} else {
				return false;
			}
		}

		public static Vector3D VectorNormalSUR(Cara Cara, Vertice[] Vertices)
		{
			return (new Vector3D(Vertices[Cara.Vertices[0]].CoodenadasSUR, Vertices[Cara.Vertices[1]].CoodenadasSUR) + new Vector3D(Vertices[Cara.Vertices[0]].CoodenadasSUR, Vertices[Cara.Vertices[2]].CoodenadasSUR)).VectorUnitario;
		}

		public static Vector3D VectorNormalSRC(Cara Cara, Vertice[] Vertices)
		{
			return (new Vector3D(Vertices[Cara.Vertices[0]].CoodenadasSRC, Vertices[Cara.Vertices[1]].CoodenadasSRC) + new Vector3D(Vertices[Cara.Vertices[0]].CoodenadasSRC, Vertices[Cara.Vertices[2]].CoodenadasSRC)).VectorUnitario;
		}

		public static Plano3D PlanoSUR(Cara Cara, Vertice[] Vertices)
		{
			return new Plano3D(Vertices[Cara.Vertices[0]].CoodenadasSUR, Vertices[Cara.Vertices[1]].CoodenadasSUR, Vertices[Cara.Vertices[2]].CoodenadasSUR);
		}

		public static Plano3D PlanoSRC(Cara Cara, Vertice[] Vertices)
		{
			return new Plano3D(Vertices[Cara.Vertices[0]].CoodenadasSRC, Vertices[Cara.Vertices[1]].CoodenadasSRC, Vertices[Cara.Vertices[2]].CoodenadasSRC);
		}

		public static Punto3D BaricentroCaraSUR(Cara Cara, Vertice[] Vertices)
		{
			return Punto3D.Baricentro(Cara.PuntosSUR[Vertices]);
		}

		public static Punto3D BaricentroCaraSRC(Cara Cara, Vertice[] Vertices)
		{
			return Punto3D.Baricentro(Cara.PuntosSRC[Vertices]);
		}

		public override string ToString()
		{
			string Retorno = "";

			for (int i = 0; i <= mVertices.GetUpperBound(0); i++) {
				if (i < mVertices.GetUpperBound(0)) {
					Retorno += mVertices[i].ToString() + ",";
				} else {
					Retorno += mVertices[i].ToString() + ",Baricentro=" + mBaricentroSUR.ToString() + ",Normal=" + mNormalSUR.ToString() + "}";
				}

			}

			return Retorno;
		}
	}
}

File: motor3d 2.0/cara.vb
Imports Motor3D.Espacio3D
Imports Motor3D.Espacio2D
Imports Motor3D.Escena
Imports Motor3D.Escena.Shading
Imports System.Drawing
Imports System.Math

Namespace Primitivas3D
    Public Class Cara
        Inherits ObjetoPrimitiva3D

        Private mNormalSUR As Vector3D
        Private mNormalSRC As Vector3D
        Private mVertices() As Integer
        Private mBaricentroSUR As Punto3D
        Private mBaricentroSRC As Punto3D
        Private mColor As Color
        Private mColorShading As Color
        Private mCargadaEnBuffer As Boolean

        Public Shadows Event Modificado(ByRef Sebder As Cara)

        Public Property CargadaEnBuffer As Boolean
            Get
                Return mCargadaEnBuffer
            End Get
            Set(value As Boolean)
                mCargadaEnBuffer = value
            End Set
        End Property

        Public Property NormalSUR() As Vector3D
            Get
                Return mNormalSUR
            End Get
            Set(ByVal value As Vector3D)
                mNormalSUR = value
                RaiseEvent Modificado(Me)
            End Set
        End Property

        Public Property NormalSRC() As Vector3D
            Get
                Return mNormalSRC
            End Get
            Set(ByVal value As Vector3D)
                mNormalSRC = value
                RaiseEvent Modificado(Me)
            End Set
        End Property

        Public Property BaricentroSUR() As Punto3D
            Get
                Return mBaricentroSUR
            End Get
            Set(ByVal value As Punto3D)
                mBaricentroSUR = value
                RaiseEvent Modificado(Me)
            End Set
        End Property

        Public Property BaricentroSRC() As Punto3D
            Get
                Return mBaricentroSRC
            End Get
            Set(ByVal value As Punto3D)
                mBaricentroSRC = value
                RaiseEvent Modificado(Me)
            End Set
        End Property

        Public Property Vertices() As Integer()
            Get
                Return mVertices
            End Get
            Set(ByVal value As Integer())
                If value.GetUpperBound(0) = mVertices.GetUpperBound(0) Then
                    mVertices = value
                    RaiseEvent Modificado(Me)
                End If
            End Set
        End Property

        Public Property Vertices(ByVal Indice As Integer) As Integer
            Get
                If Indice >= 0 AndAlso Indice <= mVertices.GetUpperBound(0) Then
                    Return mVertices(Indice)
                Else
                    Return -1
                End If
            End Get
            Set(ByVal value As Integer)
                If Indice >= 0 AndAlso Indice <= mVertices.GetUpperBound(0) Then
                    mVertices(Indice) = value
                    RaiseEvent Modificado(Me)
                End If
            End Set
        End Property

        Public ReadOnly Property Vertices(ByVal ArrayVertices() As Vertice) As Vertice()
            Get
                Dim Retorno(mVertices.GetUpperBound(0)) As Vertice

                For i As Integer = 0 To mVertices.GetUpperBound(0)
                    Retorno(i) = ArrayVertices(mVertices(i))
                Next

                Return Retorno
            End Get
        End Property

        Public ReadOnly Property PuntosSUR(ByVal Vertices() As Vertice) As Punto3D()
            Get
                Dim Retorno(mVertices.GetUpperBound(0)) As Punto3D

                For i As Integer = 0 To mVertices.GetUpperBound(0)
                    Retorno(i) = Vertices(mVertices(i)).CoodenadasSUR
                Next

                Return Retorno
            End Get
        End Property

        Public ReadOnly Property PuntosSRC(ByVal Vertices() As Vertice) As Punto3D()
            Get
                Dim Retorno(mVertices.GetUpperBound(0)) As Punto3D

                For i As Integer = 0 To mVertices.GetUpperBound(0)
                    Retorno(i) = Vertices(mVertices(i)).CoodenadasSRC
                Next

                Return Retorno
            End Get
        End Property

        Public ReadOnly Property Representacion(ByVal Vertices() As Vertice) As Punto2D()
            Get
                Dim Retorno(mVertices.GetUpperBound(0)) As Punto2D

                For i As Integer = 0 To mVertices.GetUpperBound(0)
                    Retorno(i) = Vertices(mVertices(i)).Representacion
                Next

                Return Retorno
            End Get
        End Property

        Public ReadOnly Property RepresentacionToPoint(ByVal Vertices() As Vertice) As Point()
            Get
                Dim Retorno(mVertices.GetUpperBound(0)) As Point

                For i As Integer = 0 To mVertices.GetUpperBound(0)
                    Retorno(i) = Vertices(mVertices(i)).Representacion.ToPoint
                Next

                Return Retorno
            End Get
        End Property

        Public ReadOnly Property NumeroVertices() As Integer
            Get
                Return mVertices.GetUpperBound(0) + 1
            End Get
        End Property

        Public Property Color() As Color
            Get
                Return mColor
            End Get
            Set(ByVal value As Color)
                mColor = value
                RaiseEvent Modificado(Me)
            End Set
        End Property

        Public Property ColorShading() As Color
            Get
                Return mColorShading
            End Get
            Set(ByVal value As Color)
                mColorShading = value
            End Set
        End Property

        Public Sub New(ByVal ParamArray Vertices() As Integer)
            If Vertices.GetUpperBound(0) >= 2 Then
                mVertices = Vertices
                mNormalSUR = New Vector3D
                mBaricentroSUR = New Punto3D
                mColor = Color.White
            Else
                If Vertices.GetUpperBound(0) = 0 Then
                    ReDim mVertices(Vertices(0) - 1)
                    mNormalSUR = New Vector3D
                    mBaricentroSUR = New Punto3D
                    mColor = Color.White
                Else
                    Throw New ExcepcionPrimitiva3D("CARA (NEW): Una cara debe tener al menos tres vértices." & vbNewLine _
                                                  & "Número de vértices: " & Vertices.GetUpperBound(0) + 1)
                End If

            End If
        End Sub

        Public Sub RecalcularNormalSUR(ByVal Vertices() As Vertice)
            mNormalSUR = Cara.VectorNormalSUR(Me, Vertices)
            RaiseEvent Modificado(Me)
        End Sub

        Public Sub RecalcularBaricentroSUR(ByVal Vertices() As Vertice)
            mBaricentroSUR = Cara.BaricentroCaraSUR(Me, Vertices)
            RaiseEvent Modificado(Me)
        End Sub

        Public Sub RecalcularDatosSUR(ByVal Vertices() As Vertice)
            mNormalSUR = Cara.VectorNormalSUR(Me, Vertices)
            mBaricentroSUR = Cara.BaricentroCaraSUR(Me, Vertices)
            RaiseEvent Modificado(Me)
        End Sub

        Public Sub RecalcularNormalSRC(ByVal Vertices() As Vertice)
            mNormalSRC = Cara.VectorNormalSRC(Me, Vertices)
            RaiseEvent Modificado(Me)
        End Sub

        Public Sub RecalcularBaricentroSRC(ByVal Vertices() As Vertice)
            mBaricentroSRC = Cara.BaricentroCaraSRC(Me, Vertices)
            RaiseEvent Modificado(Me)
        End Sub

        Public Sub RecalcularDatosSRC(ByVal Vertices() As Vertice)
            mNormalSRC = Cara.VectorNormalSRC(Me, Vertices)
            mBaricentroSRC = Cara.BaricentroCaraSRC(Me, Vertices)
            RaiseEvent Modificado(Me)
        End Sub

        Public Sub Shading(ByVal Constantes As PhongShader, ByVal Focos() As Foco3D, ByVal Camara As Camara3D)
            mColorShading = Constantes.EcuacionPhong(Focos, mNormalSUR, mBaricentroSUR, mColor, Camara)
        End Sub

        Public Sub RevertirVertices()
            Array.Reverse(mVertices)
        End Sub

        Public Function EsVisible(ByVal Camara As Camara3D, ByVal Vertices() As Vertice) As Boolean
            If mNormalSRC.Z <= 0.01 AndAlso Camara.Frustum.Pertenece(mBaricentroSRC) Then
                For i As Integer = 0 To mVertices.GetUpperBound(0)
                    If Camara.Pantalla.Pertenece(Vertices(mVertices(i)).Representacion) Then
                        Return True
                    End If
                Next

                Return False
            Else
                Return False
            End If
        End Function

        Public Shared Function VectorNormalSUR(ByVal Cara As Cara, ByVal Vertices() As Vertice) As Vector3D
            Return (New Vector3D(Vertices(Cara.Vertices(0)).CoodenadasSUR, Vertices(Cara.Vertices(1)).CoodenadasSUR) & New Vector3D(Vertices(Cara.Vertices(0)).CoodenadasSUR, Vertices(Cara.Vertices(2)).CoodenadasSUR)).VectorUnitario
        End Function

        Public Shared Function VectorNormalSRC(ByVal Cara As Cara, ByVal Vertices() As Vertice) As Vector3D
            Return (New Vector3D(Vertices(Cara.Vertices(0)).CoodenadasSRC, Vertices(Cara.Vertices(1)).CoodenadasSRC) & New Vector3D(Vertices(Cara.Vertices(0)).CoodenadasSRC, Vertices(Cara.Vertices(2)).CoodenadasSRC)).VectorUnitario
        End Function

        Public Shared Function PlanoSUR(ByVal Cara As Cara, ByVal Vertices() As Vertice) As Plano3D
            Return New Plano3D(Vertices(Cara.Vertices(0)).CoodenadasSUR, Vertices(Cara.Vertices(1)).CoodenadasSUR, Vertices(Cara.Vertices(2)).CoodenadasSUR)
        End Function

        Public Shared Function PlanoSRC(ByVal Cara As Cara, ByVal Vertices() As Vertice) As Plano3D
            Return New Plano3D(Vertices(Cara.Vertices(0)).CoodenadasSRC, Vertices(Cara.Vertices(1)).CoodenadasSRC, Vertices(Cara.Vertices(2)).CoodenadasSRC)
        End Function

        Public Shared Function BaricentroCaraSUR(ByVal Cara As Cara, ByVal Vertices() As Vertice) As Punto3D
            Return Punto3D.Baricentro(Cara.PuntosSUR(Vertices))
        End Function

        Public Shared Function BaricentroCaraSRC(ByVal Cara As Cara, ByVal Vertices() As Vertice) As Punto3D
            Return Punto3D.Baricentro(Cara.PuntosSRC(Vertices))
        End Function

        Public Overrides Function ToString() As String
            Dim Retorno As String = ""

            For i As Integer = 0 To mVertices.GetUpperBound(0)
                If i < mVertices.GetUpperBound(0) Then
                    Retorno &= mVertices(i).ToString & ","
                Else
                    Retorno &= mVertices(i).ToString & ",Baricentro=" & mBaricentroSUR.ToString & ",Normal=" & mNormalSUR.ToString & "}"
                End If

            Next

            Return Retorno
        End Function
    End Class
End Namespace

